clang/test/CodeGenSYCL/free_function_int_header.cpp

```diff
@@ -0,0 +1,48 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -triple spir64-unknown-unknown -sycl-std=2020 -fsycl-int-header=%t.h %s
+// RUN: FileCheck -input-file=%t.h %s
+// 
+// This test checks integration header contents for free functions with scalar
+// and pointer parameters.
+
+#include "mock_properties.hpp"
+#include "sycl.hpp"
+
+__attribute__((sycl_device))
+[[__sycl_detail__::add_ir_attributes_function("sycl-single-task-kernel",
+                                              2)]] void
+ff_2(int *ptr, int start, int end) {
+  for (int i = start; i <= end; i++)
+    ptr[i] = start + 66;
+}
+
+template <typename T>
+__attribute__((sycl_device))
+[[__sycl_detail__::add_ir_attributes_function("sycl-nd-range-kernel", 2)]] void
+ff_3(T *ptr, T start, int end) {
+  for (int i = start; i <= end; i++)
+    ptr[i] = start;
+}
+
+// Explicit instantiation with "int*"
+template void ff_3(int *ptr, int start, int end);
```

Comment:
Please add a second explicit instantiation to ensure multiple instantiations are handled correctly.

Please also add an explicit specialization to ensure that non-instantiated specializations are handled correctly.

Please exercise overloaded declarations at file scope, namespace scope, and inline namespace scope to ensure that unique names are correctly generated for each case. 

And finally, please add tests for static member functions (assuming that they are expected to work at this point; if not, I'll look forward to those tests in a future PR).

Reply:
Please also add tests for free functions in unnamed namespaces.

Are free functions allowed to be overloads of operators or user defined literals declared at namespace scope? Regardless, we should have tests for those to ensure that they either work as intended and produce meaningful kernel names (which would require a different naming strategy since there is nothing to prepend with `__sycl_kernel_`) or that an appropriate diagnostic is issued.

Reply:
Current support is for file scope free functions only.